Add RangeReplaceableCollection conformance

interstateone opened the pull request:
This wasn't as hard as I thought it might be 😄. Closes #3.

`removeAll(keepingCapacity:)` now uses the default implementation from RangeReplaceableCollection. The initializer needed to be added because the compiler couldn't choose between the two default implementations provided by SetAlgebra and RangeReplaceableCollection (I think.)

interstateone committed:
Add RangeReplaceableCollection conformance

KenjiTakahashi commented on Sources/OSet.swift:
```diff
@@ -273,6 +265,25 @@ public struct OSet<E: Hashable>: SetAlgebra, MutableCollection, RandomAccessColl
+
+    // MARK: RangeReplaceableCollection
+
+    public init<S>(_ elements: S) where S: Sequence, Self.Element == S.Element {
```
Using `Self` inside `where` clauses was made legal in 5.2, but I need to have support for the earlier versions. Please change these to explicit `OSet` references.

KenjiTakahashi reviewed (changes requested):
Cool, just as I was about to tag a new version :-).

The `init` part seems to be a kind of "protocol diamond problem" case. As far as I can tell, there is no better way of solving this. Still, I'd rather use the implementation from `RangeReplaceableCollection` (i.e. `append(contentsOf: elements)`). It is more succinct and should also be slightly faster (at least in some cases), because it preallocates needed memory beforehand.

Changing the above also accidentally unveils problem with `replaceSubrange` implementation, though. It does not work correctly when new elements contain duplicates to the ones outside replace range. Example test case: `([1, 2], 0...0, [2], [2])`.

interstateone commented:
Thanks for catching that bug! I've added that test case, as well as another to make sure that duplicate elements are properly removed and added to reflect a change in index, and changed the implementation to pass both. Let me know if there are more improvements to make there, I'm not sure about the complexity (in the big-O sense) of that implementation. I made those other two changes (init, Self) as well.

(Nice GitHub Actions matrix!)

interstateone commented:
(Oops, lazy push, will take a better look at it to get it passing again)

interstateone commented:
I've made another change that I think has the correct behaviour now, and I ran all the tests before pushing this time 🙃. I was overthinking it a bit before. I also added an implementation of `reserveCapacity` since the default is a no-op and you'd mentioned the potential benefits of its use in the initializer.

KenjiTakahashi commented on Tests/OSetTests/OSetTests.swift:
```diff
@@ -31,6 +33,43 @@ class OSetTests: XCTestCase {
+            ([1, 3, 5], 0...2, [2, 4, 5], [2, 4, 5]),
+            ([1, 3, 5], 0...2, [], []),
+            ([1, 2], 0...0, [2], [2]),
+            ([1, 2, 3], 0...0, [3], [2, 3])
```
This is interesting case, my gut feeling would be that it should move the existing item (i.e. create `[3, 2]` result). But decided to cross-check it with `[NSMutableOrderedSet -replaceObjectsAtIndexes:withObjects]` and also Python's `OrderedDict` (not the same thing, but similar) and they both keep the old element position. So I guess it's fine this way.

interstateone replied:
Yeah, I was also thinking that it should move the element at first, and I was struggling with the implementation. Not moving the element made the implementation easier. Thanks for double-checking that decision!

KenjiTakahashi reviewed (approved):
Oh, didn't realise the default `reserveCapacity` is no-op. Although it makes sense when you think about it.
Anyway, looks good to me now. Thanks!
